### PR TITLE
Allow setting active by segment

### PR DESCRIPTION
**Pull request by bencrouse**
This allows configuring releasable resources to be active only for
certain segments. If no segments are specified, it will be active
globally. If segments are specified, only those segments will be able to
see it.

For #102

### CODE REVIEW

File: admin/app/assets/javascripts/workarea/admin/modules/select2.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
+/**
```
Comment (mttdffy) [inline admin/app/assets/javascripts/workarea/admin/modules/select2.js]: im confused by the addition of this file, but I'll leave the judging to Curt

File: admin/app/assets/javascripts/workarea/admin/modules/toggle.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
+/**
```
Comment (mttdffy) [inline admin/app/assets/javascripts/workarea/admin/modules/toggle.js]: thought we have this behavior already, but again I'll defer to Curt

File: admin/app/controllers/workarea/admin/segments_controller.rb [hunk 3]
```diff
@@ -3,11 +3,12 @@ module Admin
     class SegmentsController < Admin::ApplicationController
       required_permissions :people
 
-      before_action :find_segment, except: :index
+      before_action :find_segment, except: %i(index move)
```
Comment (mttdffy) [inline admin/app/controllers/workarea/admin/segments_controller.rb]: i don't see a move action?

File: admin/config/routes.rb [hunk 4]
```diff
@@ -362,10 +362,15 @@
     resources :segments, except: [:new, :create] do
       resources :rules, except: [:show, :new, :edit], controller: 'segment_rules'
 
+      collection do
+        post :move
```
Comment (mttdffy) [inline admin/config/routes.rb]: no matching action?

File: core/app/models/workarea/content/block_draft.rb [hunk 5]
```diff
@@ -29,7 +32,9 @@ def typecast_data!
       end
 
       def to_block
-        content.blocks.build(attributes.except('content_id'))
+        result = Content::Block.instantiate(as_document.except('content_id'))
```
Comment (mttdffy) [inline core/app/models/workarea/content/block_draft.rb]: TIL `.instantiate` is a thing. Whats the significance over `.new`?

File: storefront/app/view_models/workarea/storefront/recommendations_view_model.rb [hunk 6]
```diff
@@ -34,6 +32,12 @@ def product_ids
       def result_count
         raise NotImplementedError
       end
+
+      private
+
+      def find_displayable_products(ids)
+        Catalog::Product.find_ordered(ids).select(&:active?).select(&:purchasable)
```
Comment (mttdffy) [inline storefront/app/view_models/workarea/storefront/recommendations_view_model.rb]: should use `purchasable?` as that factors in variant activeness, too.

File: admin/app/assets/javascripts/workarea/admin/modules/toggle.js (hunk 2 above)
Comment (bencrouse) [reply admin/app/assets/javascripts/workarea/admin/modules/toggle.js]: yeah, I thought so too 😞

File: admin/app/assets/javascripts/workarea/admin/modules/select2.js (hunk 1 above)
Comment (bencrouse) [reply admin/app/assets/javascripts/workarea/admin/modules/select2.js]: The idea is to allow you to make a pure instance of select2 for things like inline options like `allowClear`, etc. `remoteSelects` is written specifically for remote stuff: https://github.com/workarea-commerce/workarea/blob/master/admin/app/assets/javascripts/workarea/admin/modules/remote_selects.js

File: admin/app/controllers/workarea/admin/segments_controller.rb (hunk 3 above)
Comment (bencrouse) [reply admin/app/controllers/workarea/admin/segments_controller.rb]: Yep, cruft from when segments had priority.

File: admin/config/routes.rb (hunk 4 above)
Comment (bencrouse) [reply admin/config/routes.rb]: Yep, cruft.

File: core/app/models/workarea/content/block_draft.rb (hunk 5 above)
Comment (bencrouse) [reply core/app/models/workarea/content/block_draft.rb]: Yeah, this was a TIL I learned for me too. Basically like `new`, but assumes the attributes are raw, as-if from the DB.

Comment (bencrouse) [reply core/app/models/workarea/content/block_draft.rb]: This is important here because `active_segment_ids` are localized.

**Comment by bencrouse** [comment]
@meowsus can you take a look? I'd like to get this merged ASAP so QA can begin.

File: admin/app/assets/javascripts/workarea/admin/modules/select2.js (hunk 1 above)
Comment (meowsus) [reply admin/app/assets/javascripts/workarea/admin/modules/select2.js]: @mttdffy this is something we had originally, long, long ago, when Select2 was introduced into the Admin, but we had no real use case for it (until now) so it fell by the wayside. As part of #142 I'll better align remoteSelects & select2's APIs so this makes a bit more sense.

File: admin/app/assets/javascripts/workarea/admin/modules/toggle.js (hunk 2 above)
Comment (meowsus) [reply admin/app/assets/javascripts/workarea/admin/modules/toggle.js]: I did as well. I'll triple check as part of #142.

File: admin/app/views/workarea/admin/segments/index.html.haml [hunk 7]
```diff
@@ -6,37 +6,15 @@
       .grid__cell.grid__cell--50
         .view__heading
           = link_to "↑ #{t('workarea.admin.segments.index.dashboard_link')}", people_dashboards_path, class: 'view__dashboard-button'
-          %h1= t('workarea.admin.segments.index.page_title')
+          %h1.heading--no-margin= t('workarea.admin.segments.index.page_title')
```
Comment (meowsus) [inline admin/app/views/workarea/admin/segments/index.html.haml]: Super minor: `.heading.heading--no-margin`

File: core/lib/workarea.rb [hunk 8]
```diff
@@ -182,6 +182,14 @@ def self.define_content_block_types(&block)
     require_dependency Workarea::Core::Engine.root.join('app', 'models', 'workarea', 'content', 'block_type_definition')
     Content::BlockTypeDefinition.new.instance_eval(&block)
   end
+
+  # Use this deprecation to warn about the next minor release.
+  #
+  # @return [ActiveSupport::Deprecation]
+  #
+  def self.deprecation
+    @deprecation ||= ActiveSupport::Deprecation.new('3.6', 'Workarea')
```
Comment (meowsus) [inline core/lib/workarea.rb]: I feel like passing the version to the `deprecation` method would help us grok for deprecations more easily, much like we do for `TODO v3.6` 

```ruby
def self.deprecation(version)
  @deprecation ||= ActiveSupport::Deprecation.new(version, 'Workarea')
end
```

That way we're not always forgetting  to bump this version in this method every minor, and we can more easily `ack "deprecation('3.6')"` at the same time we `ack 'TODO 3.6'`.

Just a thought.

Comment (meowsus) [reply core/lib/workarea.rb]: Loving this standardization tho. I always find myself looking up how we do this when I need to deprecate stuff.

**Pull request merged**